### PR TITLE
Treatment parameter rework

### PR DESCRIPTION
**Pull request by mark-todd**

### CODE REVIEW

File: epioncho_ibm/state/state.py [hunk 1]
```diff
@@ -93,9 +121,19 @@ def reset_params(self, params: Params):
         Args:
             params (Params): New set of parameters
         """
+        self.numpy_bit_generator = Generator(SFC64(params.seed))
```
Comment (mkcinu) [inline epioncho_ibm/state/state.py:124]: Do you always want to reinitialise it or only when compliance % changes?

File: epioncho_ibm/state/state.py [hunk 2]
```diff
@@ -65,6 +65,34 @@ def negative_binomial_alt_interface(
     return output
 
 
+def recalculate_compliance(
```
Comment (mkcinu) [inline epioncho_ibm/state/state.py:68]: This function modifies is_compliant. When it's also returned, the inteface suggests that it doesn't modify and returns a copy (possibly modified). I'd make it return None

File: epioncho_ibm/state/state.py [hunk 3]
```diff
@@ -65,6 +65,34 @@ def negative_binomial_alt_interface(
...
 
+def recalculate_compliance(
+    is_compliant: Array.Person.Bool,
+    prev_noncompliance_rate: float,
+    new_noncompliance_rate: float,
+    people_generator: Generator,
+):
+    if prev_noncompliance_rate == new_noncompliance_rate:
+        return is_compliant
+    elif prev_noncompliance_rate > new_noncompliance_rate:
```
Comment (mkcinu) [inline epioncho_ibm/state/state.py:76]: As discussed - won't work if prev is None. You could maybe change to NaN instead? But again, care needs to be taken when doing comparisons. both < and > are false when comparing with NaN.

File: epioncho_ibm/state/state.py [hunk 4]
```diff
@@ -65,6 +65,34 @@ def negative_binomial_alt_interface(
...
+        new_draw_rate = new_noncompliance_rate / prev_noncompliance_rate
+        non_comps = len(is_compliant) - sum(is_compliant)
```
Comment (mkcinu) [inline epioncho_ibm/state/state.py:78]: `is_compliant.sum()` or `np.sum(is_compliant)` works better.

File: examples/scenario3c.json [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+        {
+            "first_year": 2026,
+            "last_year": 2040,
+            "interventions": {
+                "treatment_interval": 1,
+                "permanent_infertility": 0.9
+            }
+        }
+    ]
+}
```
Comment (mkcinu) [inline examples/scenario3c.json:25]: New lines are missing. I suggest you should change your VS code settings ;) It's a "standard" to have an empty line at the end.

File: epioncho_ibm/state/state.py (hunk 1 above)
Comment (mark-todd) [reply epioncho_ibm/state/state.py:124]: This is not a compliance related change - I just changed the order

Comment (mark-todd) [reply epioncho_ibm/state/state.py:124]: (as I needed to refer to it sooner in the code)

File: epioncho_ibm/state/state.py (hunk 2 above)
Comment (mark-todd) [reply epioncho_ibm/state/state.py:68]: Does it? Oops!

Comment (mark-todd) [reply epioncho_ibm/state/state.py:68]: I can it copy I guess

File: epioncho_ibm/state/params.py [hunk 6]
```diff
@@ -25,7 +25,7 @@ class TreatmentParams(BaseModel):
         0.345  # permenent infertility in worms due to ivermectin
     )
     min_age_of_treatment: int = 5
-
+    noncompliant_percentage: float = 0.05
```
Comment (mkcinu) [inline epioncho_ibm/state/params.py:29]: blackkkk

File: epioncho_ibm/state/people.py [hunk 7]
```diff
@@ -226,9 +227,20 @@ class People(HDF5Dataclass):
     individual_exposure: Array.Person.Float
 
     def __eq__(self, other: object) -> bool:
+        if isinstance(other, People):
+            if self.compliance is not None and other.compliance is not None:
+                compliance_equal = array_fully_equal(self.compliance, other.compliance)
+            elif self.compliance is None and other.compliance is not None:
```
Comment (mkcinu) [inline epioncho_ibm/state/people.py:233]: I'd do it like that:

```py
if (what you wrote):
   ...
else:
  compliance_equal = (self.compliance is None) == (other.compliance is None)
```

File: epioncho_ibm/state/state.py (hunk 3 above)
Comment (mark-todd) [reply epioncho_ibm/state/state.py:76]: prev cannot be none here - not in type signature

File: epioncho_ibm/state/people.py [hunk 8]
```diff
@@ -328,8 +343,12 @@ def process_deaths(
 
     def get_people_for_age_group(self, age_start: float, age_end: float) -> "People":
         rel_ages = (self.ages >= age_start) & (self.ages < age_end)
+        if self.compliance is None:
+            new_compliance = None
+        else:
+            new_compliance = self.compliance[rel_ages]
         return People(
-            compliance=self.compliance[rel_ages],
+            compliance=new_compliance,
```
Comment (mkcinu) [inline epioncho_ibm/state/people.py:351]: Or... `compliance=self.compliance[rel_ages] if self.compliance is not None else None` :P

File: epioncho_ibm/state/people.py (hunk 7 above)
Comment (mark-todd) [reply epioncho_ibm/state/people.py:234]: so

```py
if self.compliance is not None and other.compliance is not None:
    compliance_equal = array_fully_equal(self.compliance, other.compliance)
else:
    compliance_equal = (self.compliance is None) == (other.compliance is None)

Comment (mkcinu) [reply epioncho_ibm/state/people.py:234]: Yep

File: epioncho_ibm/state/state.py (hunk 2 above)
Comment (mkcinu) [reply epioncho_ibm/state/state.py:68]: I'd just return None. You don't need a copy, do you? Anyway, I think either way is better than returning + modifying.

File: epioncho_ibm/state/state.py [hunk 9]
```diff
@@ -196,11 +211,16 @@ def get_treatment_count_for_age_group(
         return self.n_treatments[start_idx:end_idx].sum()
 
     def stats(self) -> StateStats:
+        if self.people.compliance is not None:
+            mean_comp = float(np.sum(self.people.compliance)) / len(
```
Comment (mkcinu) [inline epioncho_ibm/state/state.py:215]: float(np.mean(self.people.compliance))

File: epioncho_ibm/state/state.py [hunk 10]
```diff
@@ -196,11 +211,16 @@ def get_treatment_count_for_age_group(
...
+                self.people.compliance
+            )
+        else:
+            mean_comp = 0
         return StateStats(
-            percent_compliant=float(np.sum(self.people.compliance))
-            / len(self.people.compliance),
+            percent_compliant=mean_comp,
             percent_male=float(np.sum(self.people.sex_is_male))
```
Comment (mkcinu) [inline epioncho_ibm/state/state.py:222]: Same here, although I can see that calculating percenta with np.mean may not be the most obvious thing :D But summing booleans isn't either. Well.. We know that both work, but neither is intuintive to a 5-year old

**Pull request merged**